ground/src/main/java/com/google/android/ground/persistence/sync/LocalMutationSyncWorker.kt

```diff
@@ -0,0 +1,161 @@
+      localDataStore.updateMutations(incrementRetryCounts(mutations, t)).blockingAwait()
+      Result.retry()
+    }
+  }
+
+  /**
+   * Groups mutations by user id, loads each user, applies mutations, and removes processed
+   * mutations.
+   */
+  private fun processMutations(pendingMutations: ImmutableList<Mutation>): Completable {
+    val mutationsByUserId: Map<String, List<Mutation>> = groupByUserId(pendingMutations)
+    val userIds = mutationsByUserId.keys
+    return Observable.fromIterable(userIds).flatMapCompletable { userId: String ->
+      val mutations = mutationsByUserId[userId]?.toImmutableList() ?: ImmutableList.of()
+      processMutations(mutations, userId)
+    }
+  }
+
+  /** Loads each user with specified id, applies mutations, and removes processed mutations. */
+  private fun processMutations(mutations: ImmutableList<Mutation>, userId: String): Completable {
+    return localDataStore
+      .getUser(userId)
+      .flatMapCompletable { user: User -> processMutations(mutations, user) }
+      .doOnError { Timber.d("User account removed before mutation processed") }
+      .onErrorComplete()
+  }
+
+  /** Applies mutations to remote data store. Once successful, removes them from the local db. */
+  private fun processMutations(mutations: ImmutableList<Mutation>, user: User): Completable {
+    return remoteDataStore
+      .applyMutations(mutations, user)
+      .andThen(
+        processPhotoFieldMutations(mutations)
+      ) // TODO: If the remote sync fails, reset the state to DEFAULT.
+      .andThen(localDataStore.finalizePendingMutations(mutations))
+  }
+
+  /**
+   * Filters all mutations containing submission mutations with changes to photo fields and uploads
+   * to remote storage.
+   */
+  private fun processPhotoFieldMutations(mutations: ImmutableList<Mutation>): Completable {
+    return Observable.fromIterable(mutations)
+      .filter { mutation: Mutation -> mutation is SubmissionMutation }
+      .flatMapIterable { mutation: Mutation -> (mutation as SubmissionMutation).responseDeltas }
+      .filter { (_, taskType, newResponse): ResponseDelta ->
+        taskType === Task.Type.PHOTO && newResponse.isPresent
+      }
+      .map { (_, _, newResponse): ResponseDelta -> newResponse.get().toString() }
+      .flatMapCompletable { remotePath: String ->
+        Completable.fromRunnable { photoSyncWorkManager.enqueueSyncWorker(remotePath) }
+      }
+  }
+
+  private fun groupByUserId(
+    pendingMutations: ImmutableList<Mutation>
+  ): Map<String, List<Mutation>> = pendingMutations.groupBy { it.userId }
+
+  private fun incrementRetryCounts(
+    mutations: ImmutableList<Mutation>,
+    error: Throwable
+  ): ImmutableList<Mutation> =
+    mutations.map { m: Mutation -> incrementRetryCount(m, error) }.toImmutableList()
+
+  private fun incrementRetryCount(mutation: Mutation, error: Throwable): Mutation =
+    when (mutation) {
```

Comment:
Since the branches are the same, can we drop the `when`?

Reply:
`copy()` function isn't available with `Mutation` as it is not a data class. This is the only solution I could come up with. Any suggestions?

Reply:
ah, that makes sense. I think this is fine for now then. I have a few ideas but they're probably more complicated than this approach so we can maybe investigate them later.

Reply:
👍